node/src/restarter.rs

```diff
@@ -114,21 +123,29 @@ impl NodeRestarter {
                 .network_key(&name)
                 .expect("Our key is not in the committee");
             let mut primary_network =
-                WorkerToPrimaryNetwork::new_for_single_address(network_key.to_owned(), address)
-                    .await;
+                P2pNetwork::new_for_single_address(network_key.to_owned(), address).await;
             let message = WorkerPrimaryMessage::Reconfigure(ReconfigureNotification::Shutdown);
             let primary_cancel_handle =
                 primary_network.send(network_key.to_owned(), &message).await;
 
-            let addresses = worker_cache
+            let message = PrimaryWorkerMessage::Reconfigure(ReconfigureNotification::Shutdown);
+            let mut worker_names = Vec::new();
+            for worker in worker_cache
                 .load()
                 .our_workers(&name)
                 .expect("Our key is not in the worker cache")
-                .into_iter()
-                .map(|x| x.primary_to_worker)
-                .collect();
-            let message = PrimaryWorkerMessage::Reconfigure(ReconfigureNotification::Shutdown);
-            let worker_cancel_handles = worker_network.broadcast(addresses, &message).await;
+            {
+                let address = network::multiaddr_to_address(&worker.worker_address).unwrap();
+                let peer_id = anemo::PeerId(worker.name.0.to_bytes());
+                network
+                    .connect_with_peer_id(address, peer_id)
+                    .await
```

Comment:
Is there a way this could be refactored to start all the connections in parallel instead of serially? (Or is that already how this works and I'm misunderstanding the semantics of `connect_wtih_peer_id`?)

Reply:
yeah i suppose we could, but idk if its strictly worth it, this will only be an issue the first time a connection is established and on subsequent iterations the connection will already have been established so this should result in a noop.